Привет, меня зовут Гребенюк Артём. Сегодня я буду делать ревью твоего проекта!\
В Практикуме принято общаться на "ты", однако если это неудобно - дай знать, и можем перейти на более официальное общение.

Ниже ты найдешь мои комментарии - пожалуйста, не перемещай, не изменяй и не удаляй их. Я буду использовать цветовую разметку: 

<div class="alert alert-block alert-success">
<b>Success:</b> Если все сделано хорошо. </div>

<div class="alert alert-block alert-warning">
<b>Remarks:</b> Некритичное замечание, если таких замечаний меньше трех — проект может быть принят без их отработки.</div>

<div class="alert alert-block alert-danger">
<b>Needs fixing:</b> Грубая ошибка, необходимо исправить, иначе проект не будет принят.
</div>

Давай работать над проектом в диалоге: если ты что-то меняешь в проекте или отвечаешь на мои комменатри — пиши об этом. Мне будет легче отследить изменения, если ты выделишь свои комментарии:
Например так:
<div class="alert alert-block alert-info">
<b>Fixed:</b> 
Hello World!
</div>

Или можешь использовать эту галочку ✔ в моих комментариях, если писать особо нечего.

<div class="alert alert-block alert-warning">
<b>Общее впечатление:</b> 
    
- Спасибо за работу :)
- Определенно могу сказать, что справился ты очень хорошо.
- У меня есть пара ремарок, которые помогут тебе сделать проект еще лучше.
- Пробегись по моим комментариям, исправь недочёты и всё будет отлично :)
    
</div>


<div class="alert alert-block alert-info">
<b>Fixed:</b> 
Спасибо за обратную связь, твои комментарии очень важны для меня. Очень приятно, что ты поправляешь меня в таких ньюансах, о которых я даже подумать бы и не смог :)
</div>

<div class="alert alert-block alert-success">
<b>Привет!</b> отличная работа над ремарками! Здорово, что все мои комментарии были проработаны, вне зависимости от цвета. Проект принят, поздравляю! Желаю удачи в дальнейшем обучении :) 

## Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

Исследование будет проводится в несколько этапов:
1. Анализ исходных данных;
2. Выявление аномалий в данных;
3. Замена пропущенных значений;
4. Изменение типов данных;
5. Поиск и удаление дубликатов;
6. Лемматизация;
7. Выявление зависимостей.

В результате работы будет сформирован итоговый вывод и рекомендации к построению модели.

<div class="alert alert-block alert-warning">
<b>Remarks:</b> 
    
Отлично, что у проекта есть введение. Однако стоит задуматься, как еще его можно дополнить, чтобы сделать более ценным для читателя. В студенческом ноушене (если нет доступа, спроси куратора) есть статья о том, как мы оформляем работы.
    
Также советую тебе взглянуть на <a href="https://www.kaggle.com/notebooks/">работы</a> профессиональных аналитиков, чтобы оценить, как другие люди оформляют и проводят исследования. </div>

<div class="alert alert-block alert-info">
<b>Fixed:</b> 
Поправил! Если честно, сложно писать комментарии к работе, но хотелось бы научиться. Поэтому - если есть ещё идеи, чем дополнить описание, поделись, пожалуйста :)
</div>

<div class="alert alert-block alert-success">
Мне кажется, было бы здорово вынести какие-то тезисы в начало проекта, как в научной работе мы бы сделали.

### Шаг 1. Откройте файл с данными и изучите общую информацию. 

In [1]:
import numpy as np
import pandas as pd

from pymystem3 import Mystem
from collections import Counter

In [2]:
data = pd.read_csv('/datasets/data.csv')
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

<div class="alert alert-block alert-warning">
<b>Remarks:</b> 
    
- стоит отделять импорты от открытия датасета, это делается для того, чтобы иметь возможность в любйо момент вернуться в начало, дописать нужные библиотеки, перезапустить ячейку и продолжить работу, не перетирая датасет.
- все импорты и ключевые константы рекмоендую собирать в начале проекта

<div class="alert alert-block alert-info">
<b>Fixed:</b> 
Поправил!
</div>

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [4]:
data.describe().round(3)

children  days_employed  dob_years  education_id  family_status_id  \
count  21525.000      19351.000  21525.000     21525.000         21525.000   
mean       0.539      63046.498     43.293         0.817             0.973   
std        1.382     140827.312     12.575         0.548             1.420   
min       -1.000     -18388.950      0.000         0.000             0.000   
25%        0.000      -2747.424     33.000         1.000             0.000   
50%        0.000      -1203.370     42.000         1.000             0.000   
75%        1.000       -291.096     53.000         1.000             1.000   
max       20.000     401755.400     75.000         4.000             4.000   

            debt  total_income  
count  21525.000     19351.000  
mean       0.081    167422.302  
std        0.273    102971.566  
min        0.000     20667.264  
25%        0.000    103053.153  
50%        0.000    145017.938  
75%        0.000    203435.068  
max        1.000   2265604.029

### Вывод

Очень интересный набор данных, однако в нём есть некоторые недостатки:
* Пропущенные значения в столбцах days_employed и total_income
* Отрицательные значения в столбцах children и days_employed
* Максимальное значение в столбце children составляет 20
* Минимальное значение в столбце dob_years составляет 0
* Слишком много знаков после запятой в столбцах days_employed и total_income

Ниже перечислены описания столбцов исходного файла:
* children — количество детей в семье
* days_employed — общий трудовой стаж в днях
* dob_years — возраст клиента в годах
* education — уровень образования клиента
* education_id — идентификатор уровня образования
* family_status — семейное положение
* family_status_id — идентификатор семейного положения
* gender — пол клиента
* income_type — тип занятости
* debt — имел ли задолженность по возврату кредитов
* total_income — ежемесячный доход
* purpose — цель получения кредита

<div class="alert alert-block alert-success">
<b>Success:</b> Отличный первый взгляд на данные

### Шаг 2. Предобработка данных

### Обработка пропусков

In [5]:
# Функция для поиска пропущенных значений
def finding_missing_values(data):
    missing_values = data.isnull().sum().to_frame('count')
    missing_values['percent'] = missing_values['count'] / data.shape[0] * 100
    missing_values = missing_values.round(2)
    
    return missing_values

In [6]:
finding_missing_values(data)

count  percent
children              0      0.0
days_employed      2174     10.1
dob_years             0      0.0
education             0      0.0
education_id          0      0.0
family_status         0      0.0
family_status_id      0      0.0
gender                0      0.0
income_type           0      0.0
debt                  0      0.0
total_income       2174     10.1
purpose               0      0.0

<div class="alert alert-block alert-warning">
<b>Remarks:</b> Отличная идея! Давай еще добавим в аутпут % пропусков для наглядности и организуем этот код в функцию для переиспользуемости

<div class="alert alert-block alert-info">
<b>Fixed:</b> 
Поправил! Мне кажется, что у меня не самое оптимальное решение. Если есть более красивое решение, можешь им поделиться, пожалуйста?)
</div>

<div class="alert alert-block alert-success">
Можно добавить аргумент, в зависимости от которого столбцы с 0 пропусков будут выводиться или не будут. Но в целом отличная реализация

Замена пропущенных значений:

In [7]:
data['days_employed'] = data['days_employed'].fillna(0)
data['total_income'] = data['total_income'].fillna(0)

<div class="alert alert-block alert-warning">
<b>Remarks:</b> Старайся комментарии к исследованию писать в маркдаун ячейках, а # оставить только для технических пояснений к коду

<div class="alert alert-block alert-info">
<b>Fixed:</b> 
Поправил! Спасибо, буду учитывать в будущем.
</div>

In [8]:
# Заменим отрицательные значения в столбце children на положительные
data['children'] = data['children'].abs()

# Заменим значение 20 в столбце children на 2 (скорее всего это просто опечатка)
data['children'] = data['children'].replace(20, 2)

# Проверим значения в столбце children
data['children'].value_counts().to_frame('Кол-во детей в группе')

Кол-во детей в группе
0                  14149
1                   4865
2                   2131
3                    330
4                     41
5                      9

<div class="alert alert-block alert-warning">
<b>Remarks:</b>  abs можно применить к столбцу без apply (и тем более lambda)

<div class="alert alert-block alert-info">
<b>Fixed:</b> 
Поправил! Спасибо, не знал об этом.
</div>

In [9]:
# Приведем к единому регистру все значения из столбца education
data['education'] = data['education'].str.lower()

In [10]:
# Заменим значение 0 в столбце dob_years на среднее значение в каждой группе
data['dob_years'] = data.groupby('income_type')['dob_years'].transform(lambda x: x.replace(0, int(x.mean())))

In [11]:
# Заменим отрицательные значения в столбце days_employed на положительные
data['days_employed'] = data['days_employed'].abs()

# Удалим все аномальные значения из столбца days_employed
data['days_employed'] = np.where((data['days_employed']/365) > 45, np.nan, data['days_employed'])

# Избавимся от пропущенных значений используя способ заполнения по подгруппам
data['days_employed'] = data['days_employed'].fillna(
    data.groupby('dob_years')['days_employed'].transform('mean')
)

<div class="alert alert-block alert-info">
<b>Fixed:</b> 
Тут тоже поправил abs().
</div>

### Вывод

* Мы видим одинаковое количество пропущенных значений в столбцах days_employed и total_income. Возможно это связано с тем, что данные люди никогда не работали или работали не официально. Так же это может быть связано с тем, что исходные данные были выгружены некорректно (техническая ошибка или ошибка заполнения данных). Пропущенные значения заменим на 0, так как заменять на среднее значение или медиану не совсем верно - это может повысить шанс выдачи кредита, но не повысит качество исходных данных;
* Отрицательные значения в столбцах children и days_employed были заменены на положительные;
* Все значения из столбца education были приведены к единому регистру;
* Аномальные значения в столбцах children, dob_years и days_employed так же были изменены.

### Замена типа данных

In [12]:
def memory_usage(pandas_obj):
    usage_b = pandas_obj.memory_usage(deep=True).sum()
    usage_mb = usage_b / 1024 ** 2
    return usage_mb


test_float64 = data.select_dtypes(include=['float64'])
test_int32 = test_float64.astype('int32')

test_float64_result = memory_usage(test_float64)
test_int32_result = memory_usage(test_int32)
percent = (1 - test_int32_result / test_float64_result) * 100

print('float64: {:5.2f} MB'.format(test_float64_result))
print('int32: {:7.2f} MB'.format(test_int32_result))
print('percent: {:6.2f} %'.format(percent))

float64:  0.33 MB
int32:    0.16 MB
percent:  49.98 %


In [13]:
# Приведем значения к целочисленному типу данных
data['days_employed'] = data['days_employed'].astype('int32')
data['total_income'] = data['total_income'].astype('int32')

<div class="alert alert-block alert-warning">
<b>Remarks:</b> 
    
Потенциал для эффективного изменения типов тут гораздо шире.
    
Давай я дам тебе две классные статьи, а ты попробуешь изменить типы данных на еще более подходящие и замеришь, сколько памяти в % ты сэкономишь:
    
https://pyprog.pro/data_types.html
    
https://habr.com/ru/company/ruvds/blog/442516/

<div class="alert alert-block alert-info">
<b>Fixed:</b> 
Поправил! Супер классное замечание, получил удовольствие от прочитанных статей и попытался реализовать это у себя в проекте :)
</div>

<div class="alert alert-block alert-success">
Здорово! Прикоснулись к бигдате :) 

### Вывод

В столбцах days_employed и total_income значениями после запятой можно пренебречь, поэтому они были приведены к целочисленныму типу данных. Данная операция практически не повлияет на качетсво исходных данных, однако, сделает их более наглядными.

### Обработка дубликатов

In [14]:
# Найдем количество дубликатов
data.duplicated().sum()

71

In [15]:
# Удалим дубликаты
data = data.drop_duplicates().reset_index(drop=True)

### Вывод

Все дублирующиеся строки были найдены и удалены.

### Лемматизация

In [16]:
m = Mystem() 

text = ' '.join(data['purpose'])
lemmas = m.lemmatize(text)

Counter(lemmas)

Counter({'покупка': 5897,
         ' ': 55023,
         'жилье': 4460,
         'приобретение': 461,
         'автомобиль': 4306,
         'дополнительный': 906,
         'образование': 4013,
         'сыграть': 765,
         'свадьба': 2324,
         'операция': 2604,
         'с': 2918,
         'на': 2222,
         'проведение': 768,
         'для': 1289,
         'семья': 638,
         'недвижимость': 6351,
         'коммерческий': 1311,
         'жилой': 1230,
         'строительство': 1878,
         'собственный': 635,
         'подержать': 853,
         'свой': 2230,
         'со': 627,
         'заниматься': 904,
         'сделка': 941,
         'получение': 1314,
         'высокий': 1374,
         'подержанный': 111,
         'профильный': 436,
         'сдача': 651,
         'ремонт': 607,
         '\n': 1})

In [17]:
# Лемматизация целей кредита
data['purpose_lemmas'] = data['purpose'].apply(lambda x: m.lemmatize(x))

### Вывод

Лемматизация помогла нам выявить, что самыми популярными целями получения кредита являются: недвижимость, автомобиль, образование.

### Категоризация данных

In [18]:
# Функция для категоризации по признаку наличия детей
def having_children(row):
    children = row['children']
    
    if children == 0:
        return 'нет детей'
    return 'есть дети'

# Категоризация по признаку наличия детей
data['children_category'] = data.apply(having_children, axis=1)

<div class="alert alert-block alert-warning">
<b>Remarks:</b> 
    
Можно в этой функции избавиться от последнего `else` и сразу выйти в return
    
https://medium.com/lemon-code/guard-clauses-3bc0cd96a2d3

<div class="alert alert-block alert-info">
<b>Fixed:</b> 
Да, точно, ты прав, совсем забыл про эту фишку)
</div>

In [19]:
# Посмотрим сколько человек в той или иной категории
data['children_category'].value_counts()

нет детей    14091
есть дети     7363
Name: children_category, dtype: int64

In [20]:
# Функция для категоризации по доходу
def distribution_of_income(row):
    total_income = row['total_income']
    
    if total_income <= 25000:
        return 'От 0 до 25к'
    elif total_income <= 50000:
        return 'От 25к до 50к'
    elif total_income <= 100000:
        return 'От 50к до 100к'
    elif total_income <= 200000:
        return 'От 100к до 200к'
    return 'От 200к'


# Категоризация по доходу
data['income_category'] = data.apply(distribution_of_income, axis=1)

<div class="alert alert-block alert-info">
<b>Fixed:</b> 
Так же поправил return.
</div>

In [21]:
# Посмотрим сколько человек в той или иной категории
data['income_category'].value_counts()

От 100к до 200к    9822
От 200к            5066
От 50к до 100к     4091
От 0 до 25к        2111
От 25к до 50к       364
Name: income_category, dtype: int64

<div class="alert alert-block alert-success">
<b>Success:</b> Хорошо, что посмотрели на рапределение значений по категориям

In [22]:
# Функция для категоризации по целям взятия кредита
def purpose_credit(row):
    purpose = row['purpose_lemmas']
    
    if 'автомобиль' in purpose:
        return 'Авто'
    if 'образование' in purpose:
        return 'Образование'
    if ('жилье' in purpose) or ('недвижиммость' in purpose):
        return 'Недвижимость'
    
    return 'Другое'


# Категоризация по целям взятия кредита
data['purpose_category'] = data.apply(purpose_credit, axis=1)

In [23]:
# Посмотрим сколько человек в той или иной категории
data['purpose_category'].value_counts()

Другое          8675
Недвижимость    4460
Авто            4306
Образование     4013
Name: purpose_category, dtype: int64

In [24]:
# Выведем преобразованный DataFrame
data.head()

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0           2873         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose                        purpose_lemmas  \
0               покупка жилья               [покупка,  , жилье, \n]   
1     приобретение автомобиля     [приобретение,  , автомобиль, \n]   
2               покупка жилья               [покупка,  , жилье, \n]   
3  дополнительное образование  [дополнительный,  , образование, \n]   
4             сыграть свадьбу             [сыграть,  , свадьба, \n]   

  children_category  income_category purpose_category  
0         есть дети          От 200к     Недвижимость  
1         есть дети  От 100к до 200к             Авто  
2         нет детей  От 100к до 200к     Недвижимость  
3         есть дети          От 200к      Образование  
4         нет детей  От 100к до 200к           Другое

### Вывод

Мы выполнили категоризацию по признаку наличия детей и установили, что у большей части клиентов нет детей. Так же была выполнена категоризация по доходу и целям взятия кредита, это нужно для того, чтобы ответить на вопросы ниже.

### Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [25]:
columns_names = ['Категория', 'Кол-во клиентов', 'Имеют задолженность', '% невозврата']

children_data = data.pivot_table(index=['children_category'], values='debt', aggfunc=['count', 'sum', 'mean']).reset_index()
children_data.columns = columns_names
children_data['% невозврата'] = children_data['% невозврата'] * 100 
children_data = children_data.sort_values(by='% невозврата', ascending=False)
children_data.round(3)

Категория  Кол-во клиентов  Имеют задолженность  % невозврата
0  есть дети             7363                  678         9.208
1  нет детей            14091                 1063         7.544

<div class="alert alert-block alert-success">
<b>Success:</b> Отличный аутпут!

### Вывод

У клиентов с детьми процент невозврата кредита выше относительно клиентов не имеющих детей.

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [26]:
family_status_data = data.pivot_table(index=['family_status'], values='debt', aggfunc=['count', 'sum', 'mean']).reset_index()
family_status_data.columns = columns_names
family_status_data['% невозврата'] = family_status_data['% невозврата'] * 100
family_status_data = family_status_data.sort_values(by='% невозврата', ascending=False)
family_status_data

Категория  Кол-во клиентов  Имеют задолженность  % невозврата
0  Не женат / не замужем             2810                  274      9.750890
3       гражданский брак             4151                  388      9.347145
4        женат / замужем            12339                  931      7.545182
1              в разводе             1195                   85      7.112971
2         вдовец / вдова              959                   63      6.569343

### Вывод

У клиентов, никогда не бывавших в браке процент невозврата выше относительно всех других категорий.

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [27]:
income_data = data.pivot_table(index=['income_category'], values='debt', aggfunc=['count', 'sum', 'mean']).reset_index()
income_data.columns = columns_names
income_data['% невозврата'] = income_data['% невозврата'] * 100
income_data = income_data.sort_values(by='% невозврата', ascending=False)
income_data

Категория  Кол-во клиентов  Имеют задолженность  % невозврата
1  От 100к до 200к             9822                  859      8.745673
0      От 0 до 25к             2111                  171      8.100426
4   От 50к до 100к             4091                  331      8.090931
2          От 200к             5066                  358      7.066719
3    От 25к до 50к              364                   22      6.043956

### Вывод

У клиентов с доходом от 100к до 200к процент невозврата выше относительно всех других категорий.

- Как разные цели кредита влияют на его возврат в срок?

In [28]:
purpose_data = data.pivot_table(index=['purpose_category'], values='debt', aggfunc=['count', 'sum', 'mean']).reset_index()
purpose_data.columns = columns_names
purpose_data['% невозврата'] = purpose_data['% невозврата'] * 100
purpose_data = purpose_data.sort_values(by='% невозврата', ascending=False)
purpose_data

Категория  Кол-во клиентов  Имеют задолженность  % невозврата
0          Авто             4306                  403      9.359034
3   Образование             4013                  370      9.220035
1        Другое             8675                  660      7.608069
2  Недвижимость             4460                  308      6.905830

<div class="alert alert-block alert-success">
<b>Success:</b> Очень хорошая работа по разделу

### Вывод

Самый высокий процент невозврата у кредитов, взятых на покупку автомобиля.

### Шаг 4. Общий вывод

Определенно есть зависимости между процентом невозврата кредита и наличием у клиента детей, его семейнего статуса, дохода и цели приобретения кредита. Хочется обратить пристальное внимание банка на кредиты, взятые на приобретения автомобиля, у них самый высокий процент невозрата. Так же важен социальный портрет клиента: не состоящие в браке клиента, без детей и с уровнем дохода от 100к до 200к склонны чаще остальных к невозврату кредита в срок.

Для наглядности, продемонстрируем результаты работы на цифрах. У людей, не состоящих в браке, процент невозврата 9.75, что на 29% выше, чем у людей, состоящих в браке. Для кредитов, взятых на покупку авто, процент невозврата составляет 9.35, что на 36% выше, чем процент невозврата кредитов, взятых на покупку недвижимости. Из этого можно сделать вывод о том, что клиенты, взявшие кредит на покупку недвижимости являются более устойчивыми плательщиками.

<div class="alert alert-block alert-warning">
    
<b>Remarks:</b> Общий вывод получился, к сожалению, недостаточно подробным. Подумай об этом так: общий вывод это резюме исследовательской работы, а не вывод в лабораторной и большинстов заказчиков начнут читать работу именно с него. Он должен содержать все обнаруженные инсайды, рекомендации и главное цифры

<div class="alert alert-block alert-info">
<b>Fixed:</b> 
Вывод так же поправил!
</div>

<div class="alert alert-block alert-success">
Гораздо содержательнее получилось. Всегда думай, как его можно "усилить" для того, чтобы человек, далекий от датасаенся получил максимум информации :)

### Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.